### DISCUSSION TITLE
How does openpnm calculate throat diffusivity based on adjacent pores

### DISCUSSION THREAD

**Discussion by alper-can** [PMEAL/OpenPNM#2020] [Q&A]
Hi All,

How openpnm calculates diffusion coefficient of throat when different diffusion coefficients are described in its two-neighbour pores. Let's suppose, we have two neighbour pores and their diffusion coefficients are very different. But, we need only one diffusion coefficient to calculate conductance of throat that connects these two pores. So, which diffusion coef. becomes dominant? In other words, what is the correlation to calculate throat diffusivity coefficient? Because, as far as I understand, we are not able to define a specific diffusion coefficient for throats. Openpnm draws a coefficient from pores. 

Best,

**Comment by ma-sadeghi** [comment]
By default, `openpnm` interpolates throat values, ex. if the diffusivity of pore i and j are 1 and 2, respectively, the diffusivity of the connecting throat will be 1.5. However, if you want to do your own interpolation (maybe a weighted average based on pore volume or anything else that's on your mind), then you can write your own throat diffusivity model and add it to your phase object. This way `openpnm` will fetch diffusivity values from your model. Here's an example:

```python
import numpy as np

def calc_throat_diffusivity(target,
                            pore_diffusivity="pore.diffusivity",
                            pore_volume="pore.volume"):
    network = target.network
    conns = network.conns
    D12 = target[pore_diffusivity][conns]
    V12 = target[pore_volume][conns]
    Dt = np.average(D12, axis=1, weights=V12)
    return Dt

phase.add_model("throat.diffusivity", model=calc_throat_diffusivity)
phase.regenerate_models()
```

**Reply by jgostick** [reply]
I would also add that your situation of two neighboring pores with very different diffusion coefficients is not very common. Are you by any change thinking about water in one pore and gas in the other?  If so, then there is also the partitioning coefficient between the phases that needs to be considered.  

Another thing, the diffusive conductance models calculate the g-value for each element in the conduit (pore-throat-pore), so if you have a unique value of diffusivity for the throat, the conductance models will automatically use it.

**Reply by alper-can** [reply]
Thank you very much for your answers. 

My confusing is that when I describe very different diffusion coefficients (10^-5 vs 10^-9) for neighbouring pores, the average value does not work (like (a10^-5+b10^-9)/2). I haven't described other phase yet. So, I was wondering if we would still consider partitioning coefficient. I guess I should define a new interpolation algorithm based on pore volume, as Sadeghi mentioned.

**Reply by ma-sadeghi** [reply]
Actually, I think you "should" create two phases via our `MultiPhase` object. See [this](https://github.com/PMEAL/OpenPNM/blob/dev/scripts/example_multiphase_diffusion.py) example.

**Reply by jgostick** [reply]
@alper-can, you mention that the interpolation do *not* give you the average of the two neighboring pores, but it should.  the interpoloate_data function literally computes an unweighted average.

**Reply by alper-can** [reply]
I have got it now. Thank you very much for your contributions.
